File: leetcode0872/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leaves1 = self.dfs([], root1)
        leaves2 = self.dfs([], root2)

        return leaves1 == leaves2
        
    def dfs(self, leaves, root):
        if root:
            leaves = self.dfs(leaves, root.left)
            leaves = self.dfs(leaves, root.right)

            if not root.left and not root.right:
                leaves.insert(0, root.val)

        return leaves

File: leetcode0872/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn leaf_similar(root1: Option<Rc<RefCell<TreeNode>>>, root2: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let mut leaves1: Vec<i32> = Vec::new();
        let mut leaves2: Vec<i32> = Vec::new();

        Solution::dfs(&mut leaves1, root1);
        Solution::dfs(&mut leaves2, root2);


        leaves1 == leaves2
    }

    pub fn dfs(leaves: &mut Vec<i32>, root: Option<Rc<RefCell<TreeNode>>>) {
        if let Some(rt) = root {
            let r = rt.borrow();
            if r.left.is_none() && r.right.is_none() {
                leaves.push(r.val);
                return;
            }
            Solution::dfs(leaves, r.left.clone());
            Solution::dfs(leaves, r.right.clone());
        }
    }
}